### ISSUE TITLE
Exporting house preview video failed

### ISSUE THREAD

**Issue by edalquist** [smeighan/xLights#2490]
**Describe the bug**
File > Export House Preview Video fails with `Exception caught in VideoExporter - 'VideoExporter - Error opening video codec context'` error in the log.

This happens on any sequence.

**Versions (please complete the following information):**
 - OS: OSX 10.15.7 (19H114)
 - xLights: 2021.03 installed via App Store

[xLightsProblem.zip](https://github.com/smeighan/xLights/files/5907937/xLightsProblem.zip)

**Comment by edshaken** [comment]
Similar issue: 

Exporting house preview video failed.  

2021-03-16 14:04:10,714 4588477888 log_base [DEBUG] Writing house-preview video to /Users/edbond/Desktop/Untitled.mp4.
2021-03-16 14:04:10,787 4588477888 log_base [ERROR] Exception caught in VideoExporter - 'VideoExporter - Error opening video codec context'
2021-03-16 14:04:10,853 4588477888 log_base [ERROR] DisplayError: Exporting house preview video failed.


OSX: 10.15.7
Xlights: 2021.08 via mac store

Not sure what you need - so included show folder - sequence is called layer demo

thanks

Ed

[XlightsShowFolder Bug.zip](https://github.com/smeighan/xLights/files/6149608/XlightsShowFolder.Bug.zip)

**Comment by DiddyDaveUK** [comment]
This issue on a MAC is still there in Xlights version 202112.  I don't have any other info to share but just wanted to make a note that this issue is still present.

**Comment by meastman** [comment]
I just experienced a similar issue in 2021.34 (AppStore build on macOS 12.0 Monterey on 2021 MacBook Pro). I was also able to reproduce it by building from source (master as of yesterday).

log:
```
2021-10-30 09:37:41,641 4373267840 log_base [DEBUG] Writing house-preview video to /Users/matte/Desktop/Feliz Navidad.mp4.
2021-10-30 09:37:41,756 4373267840 log_base [INFO] VideoExporter - exporting 500 x 834 video from 500 x 834
2021-10-30 09:37:41,790 4373267840 log_base [ERROR] Exception caught in VideoExporter - 'VideoExporter - error writing compressed video packet'
2021-10-30 09:37:41,805 4373267840 log_base [ERROR] DisplayError: Exporting house preview video failed.  
```

stderr:
```
[mp3 @ 0x14605fe00] Estimating duration from bitrate, this may be inaccurate
TimeDisplay - glVer:  4.1 Metal - 75.19  (Apple M1 Pro)(Apple)
Preview - glVer:  4.1 Metal - 75.19  (Apple M1 Pro)(Apple)
ModelPreview - glVer:  4.1 Metal - 75.19  (Apple M1 Pro)(Apple)
EffectGrid - glVer:  4.1 Metal - 75.19  (Apple M1 Pro)(Apple)
WaveForm - glVer:  4.1 Metal - 75.19  (Apple M1 Pro)(Apple)
[h264_videotoolbox @ 0x14681f200] Color range not set for yuv420p. Using MPEG range.
[mp4 @ 0x14684f000] pts (6400000) < dts (12799744) in stream 0
[aac @ 0x1440d1000] Qavg: 34508.078
[aac @ 0x1440d1000] 2 frames left in the queue on closing
```

I was finally able to get it working by cherry-picking the following ffmpeg commits into a branch based on either the n4.4 or n4.4.1 tag:

http://git.videolan.org/?p=ffmpeg.git;a=commitdiff;h=55d9d6767967794edcdd6e1bbd8840fc6f4e9315
> B-frames parameter is ignored in videotoolboxenc

https://git.videolan.org/?p=ffmpeg.git;a=commitdiff;h=b786bc7433dfe082441a57c1ba9ae9ea47904b78
> avcodec/videotoolboxenc: Fixes non-B-Frame encoding
>     
> Sets vtctx->has_b_frames to 0 if the VideoToolbox compression
> session will not emit B-frames (and, in consequence, no valid
> DTSs). Required for the handling of invalid DTSs in
> 'vtenc_cm_to_avpacket' (line 2018ff) to work correctly and not
> abort encoding with "DTS is invalid." when no B-frames are
> generated.

**Comment by hrc1079** [comment]
I'm using version 2021.35 and the issue still hasn't been resolved

**Comment by bshep** [comment]
I'm having the same issue 2021.38

Log filee shows:
`
2021-11-26 18:18:01,092 4494687744 log_base [DEBUG] Writing house-preview video to /Users/brucesheplan/Desktop/Rick Astley - Never Gonna Give You Up (Official Music Video)-dQw4w9WgXcQ.mp4.
2021-11-26 18:18:01,210 4494687744 log_base [ERROR] Exception caught in VideoExporter - 'VideoExporter - Error opening video codec context'
2021-11-26 18:18:01,228 4494687744 log_base [ERROR] DisplayError: Exporting house preview video failed.  
`

**Comment by meastman** [comment]
@bshep I followed the build instructions here: https://github.com/xLightsSequencer/xLights-macOS which includes building the ffmpeg libraries. Before building those, I applied the 2 patches referenced above, which @dkulp has since [modified the instructions to mention](https://github.com/xLightsSequencer/xLights-macOS/commit/f1e8465dc281d98a0d389d8244c77db2830a57c5) along with 2 others (thanks!)

I'm not sure which released versions (if any) include those patches, but the official 2021.38 release no longer gives me an error while exporting, although the video times values seem to be off by a factor of 25000; I'll file another bug after more testing.

**Comment by bshep** [comment]
> @bshep I followed the build instructions here: https://github.com/xLightsSequencer/xLights-macOS which includes building the ffmpeg libraries. Before building those, I applied the 2 patches referenced above, which @dkulp has since [modified the instructions to mention](https://github.com/xLightsSequencer/xLights-macOS/commit/f1e8465dc281d98a0d389d8244c77db2830a57c5), along with 2 other patches (thanks!)
> 
> I'm not sure which released versions (if any) include those patches, but the official 2021.38 release no longer gives me an error while exporting, although the video times seem to be off by a factor of 25000; I'll probably file another bug tonight after more testing.

Thanks for the information, for me it still doesnt work on 2021.38 (just checked again) but i dont have an M1 mac, I actually have a hackintosh, maybe some weird hardware issue... I'll try to render it on one of my windows PCs since this is not a feature I use often.

**Comment by dkulp** [comment]
You can also use the screen recorder built into OSX..    Hit Shift-Cmd-5 and use the "Record" options in the popup.

**Comment by dkulp** [comment]
Also, try resizing the preview window.   Make it much larger and try again.  I see some reports of it failing with smaller frame sizes.

**Comment by bshep** [comment]
Issue seems fixed in 2021.39
